oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/AsyncNodeStateProcessor.java

```diff
@@ -0,0 +1,140 @@
+     * subclasses.
+     *
+     * @param doc the document to process.
+     * @return whether the document should be processed or ignored.
+     */
+    protected boolean process(NodeDocument doc) {
+        return !doc.isSplitDocument();
+    }
+
+    /**
+     * Utility method that resolves he {@code uuid} into a {@link NodeState}
+     * with the help of the UUID index.
+     *
+     * @param uuid the UUID to resolve.
+     * @param resolvedPath will be set to the resolved path if available.
+     * @return the {@link NodeState} with the given UUID or {@code null} if it
+     *         cannot be resolved or doesn't exist.
+     */
+    protected final @Nullable NodeState getNodeByUUID(@NotNull String uuid,
+                                                      @NotNull AtomicReference<String> resolvedPath) {
+        PropertyState entry = uuidIndex.getChildNode(uuid).getProperty("entry");
+        NodeState state = null;
+        if (entry != null && entry.isArray() && entry.count() > 0) {
+            String path = entry.getValue(Type.STRING, 0);
+            resolvedPath.set(path);
+            state = getNode(ns.getRoot(), path);
+            if (!state.exists()) {
+                state = null;
+            }
+        }
+        return state;
+    }
+
+    @Override
+    protected final @Nullable Callable<Void> createTask(@NotNull NodeDocument document,
+                                                        @NotNull BlockingQueue<Result> results) {
+        if (process(document)) {
+            return new NodeStateTask(document, results);
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Responsibility of the subclass to implement the processor logic. This
+     * method will run as a task with an executor service.
+     *
+     * @param path the path of the {@code NodeState} to process.
+     * @param state the {@code NodeState} or {@code null} if the node does not
+     *          exist at this path. This may happen for nodes that have been
+     *          deleted but not yet garbage collected.
+     * @return the result of the task or {@code null} if nothing is reported.
+     */
+    protected abstract @Nullable Result runTask(@NotNull Path path,
+                                                @Nullable NodeState state);
+
+    protected class NodeStateTask implements Callable<Void> {
+
+        private final NodeDocument document;
+
+        private final BlockingQueue<Result> results;
+
+        public NodeStateTask(@NotNull NodeDocument document,
+                             @NotNull BlockingQueue<Result> results) {
+            this.document = document;
+            this.results = results;
+        }
+
+        @Override
+        public Void call() throws Exception {
+            Path path = document.getPath();
+            NodeState state = document.getNodeAtRevision(ns, headRevision, null);
+            Result r = runTask(path, state);
```

Comment:
This won't work because results.put() can throw an InterruptedException. I will call another method that takes a Result and handles the InterruptedException.